File: src/python/sql_injection.py
#!/usr/bin/env python3
"""
SQL Injection vulnerabilities for Python.
Used for testing SAST tools like Bandit, Checkmarx, etc.
"""

import sqlite3
import MySQLdb

# Hardcoded credentials - detectable by SAST tools
DB_USER = "admin"
DB_PASSWORD = "password123"

def vulnerable_query(user_id):
    """Vulnerable SQL query with string concatenation."""
    conn = sqlite3.connect('test.db')
    cursor = conn.cursor()

    # VULNERABLE: Direct string concatenation
    query = f"SELECT * FROM users WHERE id = '{user_id}'"
    cursor.execute(query)

    return cursor.fetchall()

def vulnerable_multiple_params(name, email):
    """SQL injection with multiple parameters."""
    conn = MySQLdb.connect(
        host="localhost",
        user=DB_USER,
        passwd=DB_PASSWORD,
        db="test"
    )
    cursor = conn.cursor()

    # VULNERABLE: Multiple parameters concatenated
    query = f"SELECT * FROM users WHERE name = '{name}' AND email = '{email}'"
    cursor.execute(query)

    return cursor.fetchall()

def vulnerable_order_by(sort_column):
    """SQL injection in ORDER BY clause."""
    conn = sqlite3.connect('test.db')
    cursor = conn.cursor()

    # VULNERABLE: User input in ORDER BY
    query = f"SELECT * FROM users ORDER BY {sort_column}"
    cursor.execute(query)

    return cursor.fetchall()

def vulnerable_second_order(user_id, order_by):
    """Multiple injection points."""
    conn = sqlite3.connect('test.db')
    cursor = conn.cursor()

    # VULNERABLE: Two injection points
    query = f"SELECT * FROM users WHERE id = '{user_id}' ORDER BY {order_by}"
    cursor.execute(query)

    return cursor.fetchall()

def safe_query(user_id):
    """Safe SQL query using parameterized statements."""
    conn = sqlite3.connect('test.db')
    cursor = conn.cursor()

    # SAFE: Using parameterized query
    query = "SELECT * FROM users WHERE id = ?"
    cursor.execute(query, (user_id,))

    return cursor.fetchall()

def exec_sqlite3(user_input):
    """Using sqlite3 module directly with vulnerable string."""
    conn = sqlite3.connect('test.db')
    cursor = conn.cursor()

    # VULNERABLE: Using executescript
    cursor.executescript(f"SELECT * FROM users WHERE id = '{user_input}'")

    return cursor.fetchall()

File: src/java/SQLInjection.java
package java;

import java.sql.*;

/**
 * This class contains SQL Injection vulnerabilities.
 * Used for testing SAST tools like Checkmarx, Fortify, SonarQube.
 */
public class SQLInjection {

    // Hardcoded credentials - detectable by most SAST tools
    private static final String DB_USER = "admin";
    private static final String DB_PASSWORD = "password123";

    /**
     * Vulnerable SQL query with string concatenation
     * SAST tools should flag this as SQL Injection
     */
    public ResultSet getUserData(String userId) throws SQLException {
        Connection conn = DriverManager.getConnection(
            "jdbc:mysql://localhost:3306/test", DB_USER, DB_PASSWORD);

        // VULNERABLE: Direct string concatenation
        String query = "SELECT * FROM users WHERE id = '" + userId + "'";
        Statement stmt = conn.createStatement();

        // VULNERABLE: Execute query with user input
        return stmt.executeQuery(query);
    }

    /**
     * Another SQL injection example with multiple parameters
     */
    public ResultSet searchUsers(String name, String email) throws SQLException {
        Connection conn = DriverManager.getConnection(
            "jdbc:mysql://localhost:3306/test", DB_USER, DB_PASSWORD);

        // VULNERABLE: Multiple parameters concatenated
        String query = "SELECT * FROM users WHERE name = '" + name
                     + "' AND email = '" + email + "'";
        Statement stmt = conn.createStatement();

        return stmt.executeQuery(query);
    }

    /**
     * SQL injection with ORDER BY clause
     */
    public ResultSet getUsersSorted(String sortColumn) throws SQLException {
        Connection conn = DriverManager.getConnection(
            "jdbc:mysql://localhost:3306/test", DB_USER, DB_PASSWORD);

        // VULNERABLE: User input in ORDER BY
        String query = "SELECT * FROM users ORDER BY " + sortColumn;
        Statement stmt = conn.createStatement();

        return stmt.executeQuery(query);
    }

    /**
     * Prepared statement example (SAFE) - for comparison
     */
    public ResultSet getUserDataSafe(String userId) throws SQLException {
        Connection conn = DriverManager.getConnection(
            "jdbc:mysql://localhost:3306/test", DB_USER, DB_PASSWORD);

        // SAFE: Using prepared statement
        String query = "SELECT * FROM users WHERE id = ?";
        PreparedStatement stmt = conn.prepareStatement(query);
        stmt.setString(1, userId);

        return stmt.executeQuery();
    }
}